dotCMS/src/main/java/com/dotmarketing/db/HibernateUtil.java

```diff
@@ -783,23 +794,149 @@ public static boolean getAsyncCommitListenersFinalization() {
+	 * called after a database transaction has been committed or the session has been closed.
+	 * This type of listener will be spawned in a new thread. This way, the main dotCMS main thread
+	 * can move on.
+	 */
+	public static class DotAsyncRunnable extends DotRunnable {
+
+		private final Runnable runnable;
+
+		public DotAsyncRunnable(final Runnable runnable) {
+			this.runnable = runnable;
+		}
+
+		@Override
+		public void run() {
+			runnable.run();
+		}
+	}
+
+	/**
+	 * Allows you to create a synchronous commit listener, which represents code that will be
+	 * called after a database transaction has been committed or the session has been closed.
+	 * Synchronous listeners will NOT RUN ON A NEW THREAD, they will use the main dotCMS thread.
+	 */
+	public static class DotSyncRunnable extends DotRunnable {
+
+		private final Runnable runnable;
+
+		public DotSyncRunnable(final Runnable runnable) {
+			this.runnable = runnable;
+		}
+
+		@Override
+		public void run() {
+			runnable.run();
+		}
+	}
+
+	/**
+	 * Adds a commit listener to the current database transaction/session. There are several
+	 * types of commit listeners, namely:
+	 * <ul>
+	 * <li>{@link DotSyncRunnable}</li>
+	 * <li>{@link DotAsyncRunnable}</li>
+	 * <li>{@link FlushCacheRunnable}</li>
+	 * <li>{@link ReindexRunnable}</li>
+	 * <li>Among others.</li>
+	 * </ul>
+	 * Commit listeners allow developers to execute code after a transaction has been committed
+	 * or the session has ended. For listeners that are not instances of {@link DotSyncRunnable}
+	 * or {@link DotAsyncRunnable} a configuration property called {@code
+	 * REINDEX_ON_SAVE_IN_SEPARATE_THREAD} determines whether listeners are executed in a
+	 * separate thread, or in the same dotCMS thread. By default, they run in a brand new thread.
+	 *
+	 * @param tag      A unique ID for the specified listener.
+	 * @param listener The commit listener wrapped as a {@link Runnable} object.
+	 *
+	 * @throws DotHibernateException An error occurred when registering the commit listener.
+	 */
+	public static void addCommitListener(final String tag, final DotRunnable listener) throws
+			DotHibernateException {
 		if (getTransactionListenersStatus() != TransactionListenerStatus.DISABLED) {
 			try {
-	    	    if(getSession().connection().getAutoCommit())
-	    	        listener.run();
-	    	    else {
-	    	    	commitListeners.get().put(tag,listener);
-	    	    }
-		    }
-		    catch(Exception ex) {
-		        throw new DotHibernateException(ex.getMessage(),ex);
-		    }
+				if(getSession().connection().getAutoCommit()) {
+					listener.run();
+				} else {
+					if (listener instanceof DotSyncRunnable) {
+						syncCommitListeners.get().put(tag, listener);
+					} else if (listener instanceof DotAsyncRunnable) {
+						asyncCommitListeners.get().put(tag, listener);
+					} else {
+						if (getAsyncCommitListenersFinalization() && Config.getBooleanProperty

```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 4 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS134)